ably/options.go

```diff
@@ -14,36 +15,86 @@ const (
 	ProtocolMsgPack = "msgpack"
 )
 
+var errInvalidKey = errors.New("invalid key format")
+
+var DefaultOptions = &ClientOptions{
+	RestEndpoint:     "rest.ably.io",
+	RealtimeEndpoint: "realtime.ably.io",
+	Protocol:         ProtocolJSON, // TODO: make it ProtocolMsgPack
+}
+
 type ClientOptions struct {
-	RealtimeEndpoint string
 	RestEndpoint     string
-	Key              string
-	Token            string
-	Secret           string
+	RealtimeEndpoint string
+	KeyName          string
+	KeySecret        string
 	ClientID         string
-	Protocol         string
+	Protocol         string // either ProtocolJSON or ProtocolMsgPack
 	UseTokenAuth     bool
 	NoTLS            bool
 
 	HTTPClient *http.Client
 }
 
-func (p *ClientOptions) Prepare() {
-	if p.Key != "" {
-		p.parseKey()
+func NewClientOptions(key string) (*ClientOptions, error) {
+	opts := &ClientOptions{}
+	if err := opts.SetKey(key); err != nil {
+		return nil, err
+	}
+	return opts, nil
+}
+
+func (opts *ClientOptions) restURL() string {
+	host := opts.RestEndpoint
+	if host == "" {
+		host = DefaultOptions.RestEndpoint
+	}
+	if opts.NoTLS {
+		return "http://" + host
+	}
+	return "https://" + host
+}
+
+func (opts *ClientOptions) realtimeURL() string {
+	host := opts.RealtimeEndpoint
+	if host == "" {
+		host = DefaultOptions.RealtimeEndpoint
+	}
+	if opts.NoTLS {
+		return "ws://" + host + ":80"
+	}
+	return "wss://" + host + ":443"
+}
+
+func (opts *ClientOptions) SetKey(key string) error {
```

Comment:
The problem the setter is solving is dependant fields: setting key field should set keyName and keySecret to proper values, However if user sets key to different value (e.g. after instantiating client but before making first request or changing the key in some handler code when the old key was wrong) the keyName and keySecret fields have still the old value, which leads to sloppy API. Thus I solved this with setter. If we have better options or we don't care the please give me feedback. 

From the description:

```
/* old API */
client := ably.NewRestClient(&ably.Options{Key: "xxx:xxx"})
// ...
client.Options.Key = "yyy:yyy"
// all calls to REST API are going to fail, as Token and Secret fields
// are still "xxx"

/* new API */
client, err := ably.NewRestClient("xxx:xxx")
// ...
err := client.Options.SetKey("yyy:yyy")
// correctly handled here
```